symengine/rational.cpp

```diff
@@ -91,4 +91,47 @@ void get_num_den(const Rational &rat,
     *num = integer((rat.i).get_num());
     *den = integer((rat.i).get_den());
 }
+
+bool Rational::is_perfect_power(bool is_expected) const
+{
+    const mpz_class &num = i.get_num();
+    if (num == 0)
+        return true;
+    else if (num == 1)
+        return mpz_perfect_power_p(i.get_den().get_mpz_t()) != 0;
+
+    const mpz_class &den = i.get_den();
+    if (den == 0)
+        return mpz_perfect_power_p(num.get_mpz_t()) != 0;
+    
+    if (not is_expected) {
+        if (mpz_cmpabs(num.get_mpz_t(), den.get_mpz_t()) > 0) {
+            if (mpz_perfect_power_p(den.get_mpz_t()) == 0)
+                return false;
+        }
+        else {
+            if (mpz_perfect_power_p(num.get_mpz_t()) == 0)
+                return false;
+        }
+    }
+    mpz_class prod = num * den;
+    return mpz_perfect_power_p(prod.get_mpz_t()) != 0;
+}
+
+bool Rational::r_nth_root(const Ptr<RCP<const Rational>> &the_rat, unsigned int n) const
+{
+    mpz_class rn;
+    int ret = mpz_root(rn.get_mpz_t(), i.get_num().get_mpz_t(), n);
+    if (ret == 0)
+        return false;
+    mpz_class rd;
+    ret = mpz_root(rd.get_mpz_t(), i.get_den().get_mpz_t(), n);
+    if (ret == 0)
+        return false;
+    mpq_class res(rn);
+    res /= rd;
+    res.canonicalize();
```

Comment:
I think it's the "Function: mpq_class::mpq_class (integer num, integer den)", since it says that `integer` is mpz_class or standard C++ integer types.